File: 234-palindrome-linked-list/palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverse(ListNode* head) {
        ListNode* curr = head, *prev = nullptr;
        while(curr) {
            auto next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
    ListNode* mid(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }
    bool isPalindrome(ListNode* head) {
        auto head_ = head;
        auto middle = mid(head_);
        auto rev = reverse(middle);
        while (rev) {
            if (rev->val != head_->val)
                return false;
            rev = rev->next;
            head_ = head_->next;
        }
        return true;
    }
};

File: 234-palindrome-linked-list/palindrome-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func half(head *ListNode) *ListNode {
    slow, fast := head, head 

    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
    }

    return slow
}

func reverse(head *ListNode) *ListNode {
    var prev *ListNode = nil 
    curr := head

    for curr != nil {
        tmp := curr
        curr = curr.Next
        tmp.Next = prev
        prev = tmp
    }  

    return prev
}

func isPalindrome(head *ListNode) bool {
    halfNode := half(head)
    reverseHalfNode := reverse(halfNode)

    for reverseHalfNode != nil {
        if head.Val != reverseHalfNode.Val {
            return false
        }

        head = head.Next
        reverseHalfNode = reverseHalfNode.Next
    }

    return true
}